File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.io.*;
import java.util.*;

public class Main {
    static class Pair{
        int y; int x;
        Pair(int y, int x){
            this.y = y;
            this.x = x;
        }
    }
    static int[] dy = {-1, 0, 1, 0}, dx = {0, 1, 0, -1};
    static int n, m;
    static int[][] board;
    static boolean[][] visited;
    static void dfs(int y, int x){
        var stk = new ArrayDeque<Pair>();
        visited[y][x] = true;
        stk.push(new Pair(y, x));
        while(!stk.isEmpty()){
            Pair p = stk.pop();
            int cy = p.y;
            int cx = p.x;
            for(int i = 0; i < 4; i++){
                int ny = cy + dy[i];
                int nx = cx + dx[i];
                if(ny < 0 || nx < 0 || ny >= n || nx >= m) continue;
                if(board[ny][nx] == 0 || visited[ny][nx]) continue;
                stk.push(new Pair(ny, nx));
                visited[ny][nx] = true;
            }
        }
    }
    public static void main(String[] args) throws IOException {
        var in = new BufferedReader(new InputStreamReader(System.in));
        var out = new PrintWriter(System.out);

        int test = Integer.parseInt(in.readLine());
        for(int t = 0; t < test; t++){
            var st = new StringTokenizer(in.readLine());
            m = Integer.parseInt(st.nextToken());
            n = Integer.parseInt(st.nextToken());
            int k = Integer.parseInt(st.nextToken());
            board = new int[n][m];
            visited = new boolean[n][m];
            for(int i = 0; i < k; i++){
                st = new StringTokenizer(in.readLine());
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());
                board[y][x] = 1;
            }
            int cnt = 0;
            for(int i = 0; i < n; i++){
                for(int j = 0; j < m; j++){
                    if(board[i][j] == 0 || visited[i][j]) continue;
                    dfs(i, j); cnt++;
                }
            }
            out.println(cnt);
        }

        out.flush();
        out.close();
    }
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
from collections import deque
t = int(input())

dy, dx = [0, 1, 0, -1], [-1, 0, 1, 0]
def dfs(y, x, vis, a):
    stk = deque()
    vis[y][x] = 1
    stk.append((y, x))
    while stk:
        cy, cx = stk.pop()
        for d in range(4):
            ny, nx = cy + dy[d], cx + dx[d]
            if ny < 0 or nx < 0 or ny >= n or nx >= m:
                continue
            if a[ny][nx] == 0 | vis[ny][nx]: continue
            vis[ny][nx] = 1
            stk.append((ny, nx))
ret = []
for _ in range(t):
    m, n, k = map(int, input().split())
    a = [[0]*m for _ in range(n)]
    vis = [[0]*m for _ in range(n)]
    for _ in range(k):
        x, y = map(int, input().split())
        a[y][x] = 1
    cnt = 0
    for i in range(n):
        for j in range(m):
            if a[i][j] and vis[i][j] == 0:
                dfs(i, j, vis, a)
                cnt += 1
    ret.append(str(cnt))
print('\n'.join(ret))